### ISSUE TITLE
Allow tmt run to execute in parallel

### ISSUE THREAD

**Issue by The-Mule** [teemtee/tmt#976]
Use case: As a tester I want to be able to run all `execute` step blocks in parallel. Things to be considered before diving into the implementation:

 * Any recommendations about the best python way to paralelize the execution?
 * How should we handle logs for individual `execute` step blocks?
 * What should the console output look like for parallel actions, including `provision` and `prepare`?

**Comment by happz** [comment]
(Unsorted notes)

> Use case: As a tester I want to be able to run all `execute` step blocks in parallel. Things to be considered before diving into the implementation:
> 
> * Any recommendations about the best python way to paralelize the execution?

Multiple shades of parallelization are available, on different levels (process pool vs thread pool) and different approaches, like a pool of workers & task queue vs spawning workers with their work already given. Threads will hit GIL sooner or later, processes eat more resources...

Also, there should be a limit on how many workers would be used (configurable...), at some point workers overwhelm their host and parallelization starts wasting resources and time, so some sort of `--max-workers` option is a must.

In our Gluetool work, we're using [futures](https://docs.python.org/3.8/library/concurrent.futures.html) with a thread-based pool of workers. This requires careful implementation of each involved piece of code, to avoid shared state as much as possible, with functions and methods stepping closer to being pure. With processes, this isolation doesn't need to be that perfect.

In both cases, some kind of "dispatcher" would need to be implemented, to split the pie into slices, with the multiprocess approach `tmt` would need to learn to accept the task as an input. But maybe that's already possible, IIRC, `tmt` can be told to run a particular test, therefore the `tmt` command I execute (`tmt run --all report`) could run all the discovery, and spawn child processes (`tmt run tests --name foo report` ?) Crude, and leaves a lot unresolved.

What about provisioning? Is `tmt` supposed to provision a machine for each slice of the pie? Or run all blocks on a single machine, at the same time? Or be given 5 machines for 20 blocks and run them on the first free machine?

I suppose answers would often be "depends", and this sounds like a call for some kind of "scheduler" (sub)component that would be open to user suggestions via options. I can only speak with some experience with our Gluetool modules, and a pipeline focusing on one particular task is easier to compose with other pipelines, therefore our modules use a dedicated scheduler that takes care of these decisions, and pipelines then focus on given (machine, preparation steps, tests to run) tuples. Whether the machine is the same as one used by another pipeline or not is out of the scope of pipelines. In TMT lands, this would be similar to this hypothetical "scheduler" that would prepare options for `provision` steps in all child pipelines, without involving `provision` implementation in the decision on whether they would share the same machine, or each of them has its own box, or whether they run on top of a pool of machines. Here's your hostname, here are your tests, don't try to peek into neighbour's garden.

> * How should we handle logs for individual `execute` step blocks?
> * What should the console output look like for parallel actions, including `provision` and `prepare`?

For real-time output, on our work downstream, we put a task-specific prefix to each and every line emitted by different workers. Easy to add, with `grep` it's trivial to get the report. Or, `tmt` could simulate what [multitail](https://linux.die.net/man/1/multitail) does, simulating "windows" for each of the workers.

For the sort of "summary" output, tmt could also emit the output from its workers, but that would require `tmt` to capture their output for later...

**Comment by lukaszachy** [comment]
So just `execute` would run in parallel? (we probably want option to cap the number of workers)
IMO if we just need to prefix output from each worker (including its start)

So for non verbose (`tmt run`)  nothing should change - progress bar will just update faster as there is no other output from execute step

For `tmt run -v`
```
    execute
        how: tmt
            [w0] 00:00:00 /script-00 [1/1]
        summary: 1 test executed
```

In log.txt e.g.
```
15:31:59             err: Warning: Permanently added '[127.0.0.1]:10040' (ED25519) to the list of known hosts.
15:32:00             [w0] test: /script-00
15:32:00             [w0] Execute '/script-00' as a 'shell' test.
15:32:00             [w0] Use workdir '/var/tmp/tmt/run-001/plans/example/discover/default/tests'.
15:32:00             [w0] Execute command 'set -eo pipefail; tmt --help' on guest '127.0.0.1'.
15:32:00             [w0] Run command 'ssh -oStrictHostKeyChecking=no -oUserKnownHostsFile=/dev/null -oIdentitiesOnly=yes -p 10040 -i /var/tmp/tmt/run-001/plans/example/provision/default/id_ecdsa root@127.0.0.1 cd '/var/tmp/tmt/run-001/plans/example/discover/default/tests'; set -eo pipefail; tmt --help'.
15:32:00             [w0] environment: None
15:32:00             [w0]     out: Warning: Permanently added '[127.0.0.1]:10040' (ED25519) to the list of known hosts.
15:32:00             [w0]     out: bash: line 1: tmt: command not found
15:32:00             [w0] Command returned '127'.
15:32:00             [w0] Append to file '/var/tmp/tmt/run-001/plans/example/execute/data/script-00/output.txt'.
15:32:00             [w0] 00:00:00 /script-00 [1/1]
15:32:00             Copy '/var/tmp/tmt/run-001/plans/example/execute/data/script-00' from the guest to '/'.
```

There is a lot of rpm / rsync operations happening: Ensuring rsync is installed and also pulling results from guest. We need to make sure they don't collide (e.g. pull rewriting other tests data)

**Comment by lukaszachy** [comment]
Good overview @happz . I read this as proposal to make just execute part parallle. (there is #267 for running whole plans in parallel).

**Comment by FrNecas** [comment]
I don't have a lot of experience with parallelism in Python, so take my words with a grain of salt. We could obviously use threads but using the [`asyncio`](https://docs.python.org/3/library/asyncio.html) library seems like the better way to go with newer versions of python. From reading up about it a bit, it seems that for each execute step, we could launch a task and then before reporting can start, we could await all of them so that the "branches join".

I see multiple problems though:
1) Output. Tmt's output is quite order-sensitive, having two parallel execute step outputs mixed up definitely wouldn't be a good thing. The same goes for logs. I am not sure, what's the correct way to approach this, maybe each execute will maintain its own output which is only shown once the whole step finishes?
2) Synchronization. As far as I understood, the idea is that there will be a synchronization beakerlib library which will handle this. So if this assumption is correct, thenn running the test is the same thing as for singlehost test, right? Simply guest.execute() the comand (e.g. the beakerlib test script). The problem I see here -- asyncio seems to run on a single core and thread, it just switches the running coroutine. The question is, will the switch happen correctly if a machine during `guest.execute()` is waiting because it's blocked by the synchronization library? The `guest.execute()` call is just a performed ssh command and if it's in process, can asyncio switch the running task?

To answer question from @lukaszachy , Provision and prepare should be parallel too, eventually, however that's more of a nice to have for multihost, whereas execute is a must have. So eventually the "flow" for multihost (2 guests in the example) should go something like this:

```
                     provision   prepare   execute                                         finish
discover          / -------\   / --------\  / --------\   gather output + report        / --------\ 
.  ------------>  .           .           .            .  --------------------------> .             .
                  \  -------/  \  -------/  \ -------/                                  \ --------/
```
and obviously `login` anywhere which would not be parallel.

Also good reminder with parallel running of complete plans. What if there were 2 plans which both were to be run in a multihost setting? That could quickly get complicated :)

**Comment by The-Mule** [comment]
@FrNecas Ad synchronization (2). Yes, synchronization is out of scope IMO and it will be handled externally. It works the same way as in Beaker - multihost test is just a singlehost one executed multiple times. 

@lukaszachy If there will be a limit for the parallelism then we need to have check that it is greater or equal than the number we need to run the test (sync library we are planning to use currently supports just a single client and a single server and hence this should be no problem, even if we use something else I don't think there will be many hosts, in most cases multihost testing is limited to 2 machines and for the rests 3 machines are usually enough). BTW: I understand that parallel execute/plans can be used and beneficial also for non-multihost testing.

**Comment by The-Mule** [comment]
@FrNecas Ad output (1), yes - I think what you are suggesting is the only way, each execute should maintain its own output and they should be printed in the end. For immediate output maybe it can work like 'tail -f output1 output2'' - where output1 and output2 are buffered and printed out in batches (triggered by time or by actual buffer size) and always prefixed. Something like:

`==> output1 <==
1
2
3
4
5

==> output2 <==
1
2
3

==> output1 <==
6
7
`

**Comment by psss** [comment]
> BTW: I understand that parallel execute/plans can be used and beneficial also for non-multihost testing.

Yes, we have several levels of where the parallel execution could be provided:

 * Run the whole plans in parallel
 * Provision & prepare pairs (we don't have to sync before prepare)
 * Test execution on individual guests (the multihost scenario)
 * Singlehost plan generating multiple environments (e.g. based on contradicting hardware requirements)
 * Run individual tests in parallel (to speed up the execution)

Looks like a bunch work to do ;-) But we can (and should) focus on one step at a time.

**Comment by sopos** [comment]
I basically agree with all what was written already. I like the idea of prefixing the immediate output if requested (`[w0] `, `[wX] `).
I would just not batch the outputs as we may still want to be able to run it in the _interactive mode_ where it is expected to run `bash` command on some of the SUTs. This would be definitely challenging, so I would consider the _interactive mode_ as a nice to have as well as **non-execute** _parallelism_, though.

**Comment by lukaszachy** [comment]
@The-Mule I think we have different interpretation of "run `execute` step in parallel". And they probably shouldn't mix.

1. UC: Speed up execute step, just for tests which can run in parallel on the _same_ guest. (similar to what pytest-xdist is doing).
2. UC: Multi-host scenarios with synchronisation between multiple guests.

**Comment by RHEmployee** [comment]
It looks like fork() and wait() scenario with some inter process communication (I think it's great to have test results before everything is finished). 

I'm not sure about another aio or threading models in python because GIL (there is work in progress to mitigate impacts of GIL).

I don't think that mixing output is something we want. Yeah, we can grep the whole pile of text and cut results to pieces but... I just don't feel it's a right way how to do it.

**Comment by kkaarreell** [comment]
As a nice to have, I would like to see the output in separate terminals. I think we could use `screen` presenting output (tail -f) of log on each system individually and a user could access them eventually.

**Comment by psss** [comment]
A related article worth reading: https://www.root.cz/clanky/soubezne-a-paralelne-bezici-ulohy-naprogramovane-v-pythonu/

**Comment by thrix** [comment]
We use [`concurrent.futures`](https://docs.python.org/3/library/concurrent.futures.html), it is [section 18.](https://www.root.cz/clanky/soubezne-a-paralelne-bezici-ulohy-programovane-v-pythonu/#k18), from the above article

**Comment by happz** [comment]
Another interesting read, https://journal.stuffwithstuff.com/2015/02/01/what-color-is-your-function/ - could be considered as advising against `async`, but I feel it's mostly about `async` being "forced into" an existing codebase. This may be a problem for TMT, but also it might be pretty safe, depending on what third-party libraries TMT uses, or how data flow through TMT plugins. I for one would avoid adopting `async` because it's perfectly possible to implement this without `async` and `async` has a tendency to spread and pollute code. I'd go for a solution based on a queue, coordinator component, and a "good ol'" worker pool crunching "jobs" - whether processes or threads, that wouldn't be that important after all (although processes are already isolated, no need to worry about changing other processes data - on the other hand, if this is supposed to work, all steps of the parallel dance would have to be clearly defined, therefore no thread should peek into its brother's garden anyway...)

In the end, the actual tool is not that important. Yes, they do have their advantages and disadvantages and will dictate boundaries of actual implementation, but the most important thing here is up to humans, to decide what should happen, model data structures, and define desired and expected behavior. I bet the actual implementation will be relatively easy once this is done.

**Comment by lukaszachy** [comment]
I've realized we might need to swap finish and report (or add 'cleanup' step). If destroying SUT stays as the very last step after all results are processed then it will just eat resources until last test/login finishes.

**Comment by happz** [comment]
As discussed yesterday, a process-based, pool-less worker approach seems to be better suited for TMT workflow than other options. Luckily, https://docs.python.org/3.7/library/multiprocessing.html does have plenty of useful primitives:

* https://docs.python.org/3.7/library/multiprocessing.html#the-process-class;
* since there's no pool of workers competing for tasks, we probably won't need `Queue` - "worker" gets its SUT input, and that's all it needs;
* workers log into their private files, therefore we're probably not interested in their stdout/stderr. We might be, to capture possible tracebacks or similar stuff, but no actual output is expected. In any case, https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Pipe does the job;
* on the other hand, our orchestrator is supposed to log interesting events in the life of its children - for example, report when a worker enters a step (e.g. "SUT ... started 'prepare'"). Some of these events can be known to orchestrators already ("SUT ... was given green light to start 'execute'"), some of them orchestrator must be told ("SUT ... sees the green light and starts 'execute'", i.e. a "confirmation" SUT got the previous "green light" message). Seems like a unidirectional stream of events, from worker to orchestrator, shall workers use their stdout? Or a dedicated https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Pipe?
* https://docs.python.org/3.7/library/multiprocessing.html#multiprocessing.Condition as a "barrier" for synchronization? `Condition` instances can be given to worker processes, our orchestrator can create several of them, one for each synchronization point (e.g. "before `execute`"), worker calls `.wait()` and then just sits until being awakened by orchestrator calling `.notify()` when the time's right.

My first impression is all tools we need to make this happen already exist and are reasonably easy to use out of the box.

**Comment by lukaszachy** [comment]
Note: we likely need to bump minimal python version (currently we are at 3.6) - Newest we could use is 3.9 (available as module for rhel/centos-stream-8)

**Comment by happz** [comment]
@lukaszachy do you see missing Python support in 3.6 for something we might need to use?

**Comment by lukaszachy** [comment]
If  we won't need kill/close from Process then there is nothing particular in mind.

**Comment by psss** [comment]
Yes, agreed, this has been covered by #1790.
